ambry-clustermap/src/main/java/com.github.ambry.clustermap/DynamicClusterChangeHandler.java

```diff
@@ -360,10 +363,20 @@ private void addOrUpdateInstanceInfos(List<InstanceConfig> instanceConfigs) thro
             // Ensure only one AmbryPartition instance exists for specific partition.
             mappedPartition = clusterChangeHandlerCallback.addPartitionIfAbsent(mappedPartition, replicaCapacity);
             ensurePartitionAbsenceOnNodeAndValidateCapacity(mappedPartition, dataNode, replicaCapacity);
-            // create new replica belonging to this partition
-            AmbryReplica replica =
-                new AmbryServerReplica(clusterMapConfig, mappedPartition, disk, stoppedReplicas.contains(partitionName),
-                    replicaCapacity, sealedReplicas.contains(partitionName));
+            // create new replica belonging to this partition or find the existing replica from bootstrapReplicas map.
+            AmbryReplica replica;
+            if (selfInstanceName.equals(instanceName)) {
+              // if this is a newly added replica on current instance, it should be present in bootstrapReplicas map.
+              replica = (AmbryReplica) bootstrapReplicas.remove(mappedPartition.toPathString());
```

Comment:
Instead of passing the map directly to DynamicClusterChangeHandler, could you add a `AmbryReplica popBootstrapReplica(String partition)` method to ClusterChangeHandlerCallback?

Reply:
Yeah, we can do that. Actually, it is quite safe in terms of concurrency. There should be only one dc (where the current node sits) calling this method.